db_stress_tool/db_stress_test_base.cc

```diff
@@ -3586,7 +3586,6 @@ void StressTest::Open(SharedState* shared, bool reopen) {
         if (inject_sync_fault || inject_open_meta_read_error ||
             inject_open_meta_write_error || inject_open_read_error ||
             inject_open_write_error) {
-          fault_fs_guard->SetInjectUnsyncedDataLoss(false);
```

Comment:
Any reason why we don't reset SetInjectUnsyncedDataLoss to false for retry reopen?

Reply:
I was thinking that it's ok for the unsycned data in a new WAL created during db open to be dropped.

Reply:
Ahh i see because we don't have any write. What about other files? It's harder to understand given we have `fault_fs_guard->SetInjectUnsyncedDataLoss(true)` but we don't have `fault_fs_guard->SetInjectUnsyncedDataLoss(false)`. Is this part (i.e, not resetting) critical to the fix?

Reply:
We only set this sync fault injection in `fault_fs_guard->SetInjectUnsyncedDataLoss(FLAGS_sync_fault_injection);` and later in https://github.com/facebook/rocksdb/blob/0bb939611dd80a2a4cb07d4fb526cc39304d9a1d/db_stress_tool/db_stress_driver.cc#L164. So resetting only affects files created during open. SST and MANIFEST are synced after writes. For WAL, our stress test set up does not support unsynced data loss when reopen is enabled, which should be covered by: https://github.com/facebook/rocksdb/blob/0bb939611dd80a2a4cb07d4fb526cc39304d9a1d/db_stress_tool/db_stress_test_base.cc#L3623-L3633.

Reply:
> Is this part (i.e, not resetting) critical to the fix?

Yes, this causes the WAL write to go to underlying file.

Reply:
I think I'm still missing something in my understanding so I think it's okay to reset un-synced data loss to be false. But I also agree with un-synced data loss = true during retrying open after injected error is fine as we don't write new data. 

So below is really for clarifying my understanding.

Here is my understanding: 
(1) First open + IO error injection + sync fault injection:  with this PR, we will create a WritableFileWriter object for 4.log (make up number) with unsync_data_loss_ = true and actual 4.log file. 
(2) First open failed 
(3) Second open + no IO error injection + no sync fault injection (to illustrate why it's okay to reset): we will create a WritableFileWriter object for 4.log (make up number) with unsync_data_loss_ = false and overwrite the actual 4.log file. Because of unsync_data_loss_== false, any write to that WAL after open will directly go into the file so we won't have the mixed buffered/direct of 4.log though this behavior of 4.log might not align with the sync fault injection settings for db run if sync_fault_injection=1.

Did I miss something?

Reply:
Ah so I initially wrote the fix that only includes removing this line. With only this line removed, it can fix the WAL corruption issue. I added the changes in fault_injection_fs later just to be future-proof. The change in fault_injection_fs should also prevent the problem as you explained.

Reply:
Got you thanks!